ISSUE 462 - store metadata as their original type instead of casting it all to string, and guess-cast it back

FJThiel opened the pull request:
This fixes #462 

#### Description
<!-- Add a high level description of the changes made (possibly quite similar to task list if it was a feature) 
e.g.
- Support new error codes
- Changed wording of some error codes
- Removed unused error codes
- Replaced some error codes that are irrelevant to user (e.g. if bouncer failed to launch, we shouldn't tell the user about it, it should just tell them the process failed)
- Add bouncer error code to email notification for better referencing.
- Changed some response codes from 500 to 400 (e.g. user uploading a file with no mesh is a user error, not system's)
-->
At the moment the metadata constructor takes in strings for the values. Then use boost to guess their original types and cast them back.

This is a bit backwards as very often we're [casting the original type to string](https://github.com/3drepo/3drepobouncer/blob/master/bouncer/src/repo/manipulator/modelconvertor/import/repo_model_import_3drepo.cpp#L63) in order to fit the function definition here, only to cast it back to the [original type](https://github.com/3drepo/3drepobouncer/blob/master/bouncer/src/repo/core/model/bson/repo_bson_factory.cpp#L223):

To correct this, a new datatype (typedef of boost::variant) was introduced that can hold all the possible data types. The whole process of processing metadata and adding it to the database was adapted to use this type. Note that formats such as Bim, DGN, and SPM still use strings for everything, since they arrive in the bouncer in that format. With the other formats, the native types are assigned to the variant. For Assimp, NWD, and Revit, helper methods for the conversion have been added.


#### Test cases
<!-- Test cases that this pull request expect to pass -->
35 new unit tests were added to test the metadata variant itself and the three conversion helper methods for Assimp, NWD, and Revit.

carmenfan reviewed (changes requested):
@FJThiel autotest failed to compile this

https://app.travis-ci.com/github/3drepo/3drepobouncer/jobs/625536714

carmenfan commented on bouncer/src/repo/manipulator/modelconvertor/import/odaHelper/data_processor_nwd.h:
```diff
@@ -42,6 +47,23 @@ namespace repo {
 					GeometryCollector* collector;
 				};
+
+				class NwdFilePathRemovalVisitor : public boost::static_visitor<> {
```
this seems very over engineered for file path removal?

it's unclear to me why we need this. Can't we sanitise the string before we create the variant?

FJThiel replied:
Mostly because that's how it was before. Not with the visitor, but cycling over all metadata at the end. Maybe because there is more than one way for a string such as that to enter the metadata. As a minimum, I would need to add  checks for the two labels and sanitation to tryConvert for the property attributes and then to setMetadataValue(std::string, OdString, OdString,...).

With tryConvert moving into the class, this is a possibility, but I see the appeal of a single location for sanitation, though it is not the option with the best performance.

carmenfan replied:
considering we know exactly which one we need to santise i feel like it'll make more sense if we do it before we put it into the metadata collection 
![image](https://github.com/user-attachments/assets/417a35cf-1582-4b7d-9826-74a9e2d9f05c)

Could wait for @sebjf's take on it if you like!

FJThiel replied:
No it's alright. I don't think this is large enough to block this until he is back.

FJThiel replied:
Done in [810b669](https://github.com/3drepo/3drepobouncer/pull/694/commits/810b669513f4efda87b17ba94a6fd6295959cd92) and [4f7ca91](https://github.com/3drepo/3drepobouncer/pull/694/commits/4f7ca91472201b248e0148f1de91cdb661597404)

sebjf replied:
For completeness, I agree that this should be before the variant creation, if only because only strings can represent paths!

FYI that `NwdFilePathRemovalVisitor` is still in the header as of the latest push, I think you intended to remove that @FJThiel!

Finally, I would rename `checkForPathAndSanitise` (e.g. to just `sanitiseString`), as currently it makes it sound like its guessing whether a string is a path and sanitising it if so. We can leave the nature of the sanitisation, including any filtering, as implementation detail of the function (and also add additional sanitisation there in the future if needed).

carmenfan commented on bouncer/src/repo/manipulator/modelconvertor/import/odaHelper/data_processor_rvt.cpp:
```diff
@@ -430,26 +332,35 @@ void DataProcessorRvt::processParameter(
+			repo::lib::MetadataVariant v;
+
+
+			//std::string variantValue = translateMetadataValue(value, labelUtils, pDescParam, element->getDatabase(), buildInEnum);
```
remove if no longer needed

FJThiel replied:
Done [3f9d159](https://github.com/3drepo/3drepobouncer/pull/694/commits/3f9d159719e63b23dd2f1f581ea14560f9a163c6)

FJThiel commented on bouncer/src/repo/core/model/bson/repo_bson_builder.h:
```diff
@@ -69,10 +74,16 @@ namespace repo {
 				}
 
+				
+				void appendMetadataVariant(
```
Do we have purposes for the type outside of metadata that requires this exact combination of storage types?

If not, what is the benefit of reusing it? As potential downsides I see that it might take types that later purposes do not (or should not) allow and that any addition with another type later on for other purposes will mean that all visitors will need to be extended, even though they are not related to the purpose of the new type.

Admittedly, these are relatively small downsides, but making a new variant type for a new purpose is just another typedef definition, which is hardly much effort either.

carmenfan replied:
The intention is to use this  very same object type when we do #1 where we will no longer piggy-back off the mongo bson object

It wouldn't make sense to have such purpose built variant type and then setup something near identical again.

FJThiel replied:
Ah, I see. I will get to it then!

FJThiel replied:
Done. [3ef77d9](https://github.com/3drepo/3drepobouncer/pull/694/commits/3ef77d9fe126f47ef1c87449d029b1eaa19209fb)

FJThiel commented on bouncer/src/repo/core/model/bson/repo_bson_builder.h:
```diff
@@ -185,6 +217,43 @@ namespace repo {
 			};
 
+			// Visitor class to process the metadata variant correctly
+			class AppendVisitor : public boost::static_visitor<> {
```
Good point. I moved it in [a390edd](https://github.com/3drepo/3drepobouncer/pull/694/commits/a390eddf31eee913f875cd1bb4a6a97bd8cacc16)

FJThiel commented on bouncer/src/repo/core/model/bson/repo_bson_factory.cpp:
```diff
@@ -220,32 +220,16 @@ MetadataNode RepoBSONFactory::makeMetaDataNode(
+		repo::lib::MetadataVariant value = entry.second;
 
-		if (!key.empty() && !value.empty())
+		if (!key.empty()) // && !value.empty())
```
Done, [cb78de9](https://github.com/3drepo/3drepobouncer/pull/694/commits/cb78de92b2a2a4b35d31ab173a6c124d64369d33)

FJThiel replied:
Correction: [3f9d159](https://github.com/3drepo/3drepobouncer/pull/694/commits/3f9d159719e63b23dd2f1f581ea14560f9a163c6)

FJThiel commented on bouncer/src/repo/manipulator/modelconvertor/import/odaHelper/data_processor_nwd.cpp:
```diff
@@ -447,32 +377,32 @@ void processAttributes(OdNwModelItemPtr modelItemPtr, RepoNwTraversalContext con
 		if (!guidAttribute.isNull())
 		{
-			setMetadataValue(category, guidAttribute->getDisplayName(), convertToStdString(guidAttribute->getGuid().toString()), metadata);
+			setMetadataValue(category, convertToStdString(guidAttribute->getDisplayName()), convertToStdString(guidAttribute->getGuid().toString()), metadata);
```
Done [553308d](https://github.com/3drepo/3drepobouncer/pull/694/commits/553308d4661524f6142293ee5b4dde8b06743d19)

FJThiel commented on bouncer/src/repo/lib/datastructure/repo_metadataVariantHelper.cpp:
```diff
@@ -0,0 +1,293 @@
+
+using namespace repo::lib;
+
+bool MetadataVariantHelper::TryConvert(aiMetadataEntry& assimpMetaEntry, MetadataVariant& v)
```
Done. The TryConvert functions were moved to their respective importers and I have applied the same split to the unit tests. [106fcd3](https://github.com/3drepo/3drepobouncer/pull/694/commits/106fcd384b1deb63fb6c6d313a9bc3ac0ec140ad)

FJThiel commented on bouncer/src/repo/lib/datastructure/repo_metadataVariant.h:
Done. [3ef77d9](https://github.com/3drepo/3drepobouncer/pull/694/commits/3ef77d9fe126f47ef1c87449d029b1eaa19209fb)

sebjf reviewed (changes requested):
Looks great, @FJThiel - neat & sensible design.

I have made a few minor comments about formatting/where things should live.

The one thing I was left wondering (and this is more for @carmenfan!) was whether for the various switch/cascaded if's, whether we should throw an exception or report an error if the case is not handled rather than continue silently, which is the current behaviour.

Currently this PR handles all possible values of the various ODA enums, so this should never be the case.

However, if ODA adds a new entry in the future it will skip that value silently.

My instinct is that this is not a problem, because like with previous missing metadata if it's important a customer will ask about it, and on the other hand making it throw will break bouncer for that model, meaning we've made it urgent for ourselves to fix.

Perhaps we should print an error though, so it's easy to see from the logs whats happened if we do get a report.

sebjf commented:
@FJThiel, as I am sure you've seen there a few failing tests as well. These two...

[  FAILED  ] RepoClientTest.UploadTestRVTRegressionTests
[  FAILED  ] RepoBSONFactoryTest.MakeMetaDataNodeTest

... will almost certainly be because the test itself assumes the type of the thing its comparing to.

(I figure you've worked out Travis by now but let me know if you want a quick call!)

FJThiel commented:
@sebjf Thanks for the feedback.

Re Travis: Yes, I am working on it since yesterday. Its gcc is certainly more picky than my local vcc so there was a lot of stuff there to do to so I only made it to the tests now.

carmenfan commented:
> The one thing I was left wondering (and this is more for @carmenfan!) was whether for the various switch/cascaded if's, whether we should throw an exception or report an error if the case is not handled rather than continue silently, which is the current behaviour.

@sebjf  yes I think handle it gracefully with a logging message should suffice

FJThiel commented on bouncer/src/repo/manipulator/modelconvertor/import/odaHelper/data_processor_rvt.h:
```diff
@@ -141,6 +137,8 @@ namespace repo {
 					OdBmSampleLabelUtilsPE* labelUtils = nullptr;
 				};
+
+				bool TryConvertMetadataEntry(OdTfVariant& metaEntry, OdBmLabelUtilsPEPtr labelUtils, OdBmParamDefPtr paramDef, OdBmDatabase* database, OdBm::BuiltInParameter::Enum param, repo::lib::RepoVariant& v);
```
This was a bit of an headache. I wanted to keep it consistent between the importers, but also consistent with the existing style. I would also have preferred to make them inaccessible from the outside, but that makes testing harder of course. So I originally settled on these global functions, since they were as unbound to a class as most other functions belonging to the data processors but, because they were global, still accessible for testing.

The Shims were a good idea because it would allow me to keep them private, but not workable with NWD because the functions that use the converter are all unbound so they can't access private or protected members of it, meaning that it would need to be public again, negating the benefit of the Shim.

In the end, I have now settled on public statics. Consistent between the processors, available for testing, and still bound to a class. Probably the best compromise.

[cc39d9b](https://github.com/3drepo/3drepobouncer/pull/694/commits/cc39d9b39e5ecdac8cff47d0ab1e902b48fc1210)

FJThiel replied:
Also, database has been removed:
[2db1b1b](https://github.com/3drepo/3drepobouncer/pull/694/commits/2db1b1bb73354dc28c307cf3340ff0551bb2a141)

FJThiel commented on bouncer/src/repo/manipulator/modelconvertor/import/repo_metadata_import_csv.h:
```diff
@@ -64,6 +65,8 @@ namespace repo{
 					std::vector<std::string> &tokenizedLine);
 				char delimiter;
+
+				std::vector<repo::lib::RepoVariant> ConvertToVariants(std::vector<std::string> tokens);
```
Done [c821bd8](https://github.com/3drepo/3drepobouncer/pull/694/commits/c821bd8aba902f098a1b0f12ce95e83a6c936de5)

FJThiel commented on bouncer/src/repo/manipulator/modelconvertor/import/odaHelper/data_processor_nwd.cpp:
```diff
@@ -447,17 +490,17 @@ void processAttributes(OdNwModelItemPtr modelItemPtr, RepoNwTraversalContext con
 		if (!guidAttribute.isNull())
 		{
-			setMetadataValue(category, guidAttribute->getDisplayName(), convertToStdString(guidAttribute->getGuid().toString()), metadata);
+			setMetadataValue(category,guidAttribute->getDisplayName(), guidAttribute->getGuid().toString(), metadata);
```
Fixed [aa7ea95](https://github.com/3drepo/3drepobouncer/pull/694/commits/aa7ea956b9710868abaef54ea66a918f1adf2d4c)

FJThiel commented on bouncer/src/repo/manipulator/modelconvertor/import/odaHelper/data_processor_rvt.cpp:
```diff
@@ -42,100 +42,63 @@ const std::set<std::string> IGNORE_PARAMS = {
-	break;
-	case OdVariant::kAnsiString:
-		strOut = std::string(val.getAnsiString().c_str());
+						   break;
```
Done [6952919](https://github.com/3drepo/3drepobouncer/pull/694/commits/69529195185a5954c3adcecc8bfc9ae2c7a8d1ba)

FJThiel commented on bouncer/src/repo/manipulator/modelconvertor/import/odaHelper/data_processor_rvt.cpp:
```diff
@@ -42,100 +42,63 @@ const std::set<std::string> IGNORE_PARAMS = {
-	case OdVariant::kString:
-		strOut = convertToStdString(val.getString());
+	switch (metaEntry.type()) {
+	case OdVariant::kVoid: {
```
Done [6952919](https://github.com/3drepo/3drepobouncer/pull/694/commits/69529195185a5954c3adcecc8bfc9ae2c7a8d1ba)

FJThiel commented on bouncer/src/repo/manipulator/modelconvertor/import/repo_model_import_assimp.cpp:
```diff
@@ -36,6 +36,55 @@
 using namespace repo::manipulator::modelconvertor;
 
+
+bool repo::manipulator::modelconvertor::TryConvertMetadataEntry(aiMetadataEntry& assimpMetaEntry, repo::lib::RepoVariant& v){
```
Done. See explanation above.

FJThiel commented:
@sebjf @carmenfan 

All feedback addressed and Travis is green.

However, you might want to check out what I did in [058d67a](https://github.com/3drepo/3drepobouncer/pull/694/commits/058d67a0037bf8b6771955e0f741f7c726c0d707). The reason the licence check Travis job would fail was that it tried to compile the conversion unit tests for rvt and nwd. However, they have ODA dependencies that are not included in that job. I moved the affected test into their own folder and added it as an exception the cmake so they would not be compiled without ODA. I hope that's alright for you.

carmenfan commented:
thanks @FJThiel 

For future reference, please click on the buttons on the top right to re-request for a review:
![image](https://github.com/user-attachments/assets/3e6b1685-0e9f-4602-94fc-139976a1c386)
That'll check the status of on the sprint board and the task will show up on the assigned person's task list!


There's also a conflict with staging, you may need to merge the latest of staging in and rerun the python script! 😄 
![image](https://github.com/user-attachments/assets/fa179c4e-d72f-4091-932e-c15999ccee48)

sebjf reviewed (approved):
I fixed a couple of case convention issues and updated the asset_generator submodule (I think its been point to issue 3 for quite a while!).

If this passes the automated tests it's good to go now.

sebjf commented:
Hi @carmenfan, while re-reviewing I put a couple of methods into camelCase & also updated where the asset_generator submodule is pointing, as its been to issue_3 all this time! No other changes though so up to you if you want to give it a second look!
